src/neuroconv/tools/roiextractors/roiextractors.py

```diff
@@ -371,6 +389,147 @@ def image_mask_iterator():
+                     traces_to_add.items() if trace is not None}
+    # Filter all zero data
+    traces_to_add = {trace_name: trace for trace_name, trace in
+                     traces_to_add.items() if any(x != 0 for x in np.ravel(trace))}
+
+    # Early return if there is nothing to add
+    if not traces_to_add:
+        return nwbfile
+
+    ophys = get_module(nwbfile, "ophys")
+
+    add_plane_segmentation(segmentation_extractor=segmentation_extractor, nwbfile=nwbfile, metadata=metadata)
+
+    # Get plane segmentation from the image segmentation
+    image_segmentation_metadata = metadata_copy["Ophys"]["ImageSegmentation"]
+    image_segmentation_name = image_segmentation_metadata["name"]
+    image_segmentation = ophys.get_data_interface(image_segmentation_name)
+
+    plane_segmentation_name = image_segmentation_metadata["plane_segmentations"][0]["name"]
+    plane_segmentation = image_segmentation.plane_segmentations[plane_segmentation_name]
+
+    # Create a reference for ROIs from the plane segmentation
+    roi_table_region = plane_segmentation.create_roi_table_region(
+        region=segmentation_extractor.get_roi_ids(),
+        description=f"region for Imaging plane{plane_index}",
+    )
+
+    rate = (
+        np.nan
+        if segmentation_extractor.get_sampling_frequency() is None
+        else float(segmentation_extractor.get_sampling_frequency())
+    )
+
+    roi_response_series_kwargs = dict(
+        rois=roi_table_region,
+        # TODO: timestamps or rate (for timestamps we need frame_to_time method)
+        # TODO: check for regularity, only use timestamps if irregular timing
+        starting_time=0.0,
+        rate=rate,
+        unit="n.a.",
+    )
+
+    for trace_name, trace in traces_to_add.items():
+        trace_name = "RoiResponseSeries" if trace_name == "raw" else trace_name.capitalize()
+        trace_name = trace_name if plane_index == 0 else trace_name + f"_Plane{plane_index}"
+
+        roi_response_series_kwargs.update(
+            name=trace_name,
+            data=np.array(trace).T,
+            rois=roi_table_region,
+
+        )
+
+        if "Dff" in trace_name:
+            # Create df/F container
+            if dff_name in ophys.data_interfaces:
+                dff = ophys.get_data_interface(dff_name)
+            else:
+                dff = DfOverF(name=dff_name)
+                # Add df/F container to the ophys module
+                ophys.add(dff)
+
+            if trace_name not in dff.roi_response_series:
+                response_series_metadata = dff_metadata["roi_response_series"]
+                trace_meta = [trace_metadata for trace_metadata in
+                              response_series_metadata if
+                              trace_name in trace_metadata["name"]]
+                if trace_meta:
+                    roi_response_series_kwargs.update(**trace_meta[0])
+
+                dff.create_roi_response_series(**roi_response_series_kwargs)
+
```

Comment:
@h-mayorquin I'd need some feedback, I understand that this is not ideal this way, and I should still move some code around because the method is called "add_fluorescence_traces" and I'm doing the a lot more here. May I ask for your suggestion how to reduce this duplication?

Reply:
The first if/else block is similar in structure to our nwb_helper function for getting modules, maybe something similar for getting NWB data interfaces could help with that?

Reply:
Otherwise another approach to flattening things like this is to add an early return statement of opposite logic, such as

```
if trace_name in dff.reaponse_series:
    continue  # continue if iteration, return if function 
```

and then all the logic of the positive case can follow in the flattened outer level

Reply:
Hi, sorry for not coming to this before after you pined me. 
My suggestion to eliminate the duplicate code and make the control structure flatter is twofold:

1) Take the logic for extracting the containers out of the loop over the traces.
2) Use the ternary operator (can also be done with default dicts) to assign metadata and containers correctly.

Code:
```python
    # Logic for extracting the containers
    if "Dff" in traces_to_add.keys():
        # Create df/F container
        if dff_name in ophys.data_interfaces:
            dff = ophys.get_data_interface(dff_name)
        else:
            dff = DfOverF(name=dff_name)
            # Add df/F container to the ophys module
            ophys.add(dff)

    remaining_traces = [trace_name for trace_name in traces_to_add.keys() if trace_name != "Dff"]
    if remaining_traces:
        # Create fluorescence container
        if fluorescence_name in ophys.data_interfaces:
            fluorescence = ophys.get_data_interface(fluorescence_name)
        else:
            fluorescence = Fluorescence(name=fluorescence_name)
            # Add fluorescence container to the ophys module
            ophys.add(fluorescence) 

    for trace_name, trace in traces_to_add.items():

        # Extract the response series metadata
        trace_name = "RoiResponseSeries" if trace_name == "raw" else trace_name.capitalize()
        trace_name = trace_name if plane_index == 0 else trace_name + f"_Plane{plane_index}"
        
        metadata = dff_metadata if trace_name == "Dff" else fluorescence_metadata
        response_series_metadata = metadata["roi_response_series"] 
        trace_metadata = next(trace_metadata for trace_metadata in response_series_metadata if trace_name==trace_metadata["name"])    
        
        # Build the roi response series
        roi_response_series_kwargs.update(**trace_metadata)
        roi_response_series = RoiResponseSeries(**roi_response_series_kwargs)
        
        # Add it to the container
        container = dff if trace_name == "Dff" else fluorescence
        container.add_roi_response_series(roi_response_series)
```

Reply:
Thank you for the suggestions @CodyCBakerPhD and @h-mayorquin , I wonder how it can be done with `defaultdicts`?

Reply:
Something like this:
```python
from collections.import defaultdict

trace_name_to_container = defaultdict(lambda: fluorescence, "Dff"=dff)
...
...
...
container = trace_name_to_container[trace_name]
container.add_roi_response_series(roi_response_series)
```

Reply:
@h-mayorquin I'm unable to get that example code to run

Reply:
They were not meant to run, just to illustrate the idea. What is it failing?

Reply:
The `defaultdict(lambda: fluorescence, "Dff"=dff)` part. `SyntaxError: expression cannot contain assignment, perhaps you meant "=="?`

Reply:
Ah, what's wrong is the string in "Dff", working example:
```python
In [1]: from collections import defaultdict

In [2]: x = defaultdict(lambda: "default_case", defined_key="defined_value")

In [3]: x["random"]
Out[3]: 'default_case'

In [4]: x["defined_key"]
Out[4]: 'defined_value'

```

Reply:
Gotcha, thanks

Reply:
thanks @h-mayorquin